internals/overlord/logstate/gatherer.go

```diff
@@ -132,18 +138,40 @@ func fillDefaultOptions(options *logGathererOptions) *logGathererOptions {
 
 // PlanChanged is called by the LogManager when the plan is changed, if this
 // gatherer's target exists in the new plan.
-func (g *logGatherer) PlanChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+// The labels map should not be modified while this method is running.
+func (g *logGatherer) PlanChanged(pl *plan.Plan, serviceData map[string]*ServiceData, labels map[string]map[string]string) {
+	// New plan - the labels may have changed. We should tell the main loop to
+	// flush the client now so that we don't later get out of sync, and send old
+	// logs with the new labels.
+	select {
+	// Check just in case the gatherer has already been stopped, so we don't get
+	// deadlocked here.
+	case <-g.tomb.Dying():
+		return
+	case g.flushCh <- struct{}{}:
+	}
+
 	// Remove old pullers
 	for _, svcName := range g.pullers.Services() {
 		svc, svcExists := pl.Services[svcName]
 		if !svcExists {
 			g.pullers.Remove(svcName)
+			select {
+			case <-g.tomb.Dying():
+				return
+			case g.setLabels <- svcWithLabels{svcName, nil}:
+			}
 			continue
 		}
 
 		tgt := pl.LogTargets[g.targetName]
 		if !svc.LogsTo(tgt) {
 			g.pullers.Remove(svcName)
+			select {
+			case <-g.tomb.Dying():
+				return
+			case g.setLabels <- svcWithLabels{svcName, nil}:
+			}
```

Comment:
These two branches above are exactly the same, and then there's pretty much nothing else. Shouldn't be hard to simplify this into a single branch.